### PR TITLE
feat: implement --build_python_zip pex

### PR DESCRIPTION
**Pull request by thesayyn**
### Type of change

- New feature or functionality (#236)

### Test plan

- New test cases added

### CODE REVIEW

File: MODULE.bazel [hunk 1]
```diff
@@ -8,7 +8,7 @@ module(
 
 # Lower-bound versions of direct dependencies.
 # When bumping, add a comment explaining what's required from the newer release.
-bazel_dep(name = "aspect_bazel_lib", version = "1.40.0")
+bazel_dep(name = "aspect_bazel_lib", version = "1.42.2")
```
Comment (thesayyn) [inline MODULE.bazel:11]: This is needed to get tar toolchain.

**Comment by thesayyn** [comment]
This is kind of working now, 

Missing things i can think of;

- inject env provided via `ctx.attr.env`
- set runfiles directory so the runfiles lookup works.
- bzlmod runfiles mapping is not working. 
- only workspace relative imports (`import wksp.xyz`) and third-party (`import 3p.xyz`) imports work, we need to find a way to inject more entries into sys.path without using venv.

File: py/private/py_pex.bzl [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    workspace_name = str(ctx.workspace_name)
+
+    args.add_all(
+        srcs, 
+        map_each = lambda f: _map_srcs(f, workspace_name),
+        uniquify = True,
+        allow_closure = True,
+    )
+    args.add(ctx.file.main, format = "--executable=%s")
+    args.add("#!/usr/bin/env python3", format="--python-shebang=%s")
```
Comment (dizzy57) [inline py/private/py_pex.bzl:37]:
```suggestion
    args.add("/usr/bin/env python3", format="--python-shebang=%s")
```

nit: `--python-shebang`'s help line says to pass shebang without `#!` (it works in both cases, whether you pass it or not)

File: py/private/py_binary.bzl [hunk 3]
```diff
@@ -6,13 +6,15 @@ load("@aspect_bazel_lib//lib:expand_make_vars.bzl", "expand_locations", "expand_
 load("//py/private:py_library.bzl", _py_library = "py_library_utils")
 load("//py/private:py_semantics.bzl", _py_semantics = "semantics")
 load("//py/private/toolchain:types.bzl", "PY_TOOLCHAIN", "VENV_TOOLCHAIN")
+load("//py/private:py_pex.bzl", "build_pex")
```
Comment (mattem) [inline py/private/py_binary.bzl:9]: Can the convention to rename on import be followed for this? _build_pex.

File: py/tools/pex/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,9 @@
+load("@rules_python//python:defs.bzl", "py_binary")
+
+py_binary(
+    name = "pex",
+    srcs = ["main.py"],
+    main = "main.py",
+    deps = ["@pypi_pex//:pkg"],
```
Comment (mattem) [inline py/tools/pex/BUILD.bazel:7]: Where does the user get this dependency from? 

rules_py currently doesn't have any Python dependencies, this will be the first, so there is no call to a pip install in the setup. This is working here as it's using the internal test time dependencies that rules_py uses.

Ideally we wouldn't have a Python dependency at all, is there another way to interact with the PEX library, can we vendor the dependency?

**Review by mattem — CHANGES_REQUESTED** [review]

File: py/tools/pex/BUILD.bazel (hunk 4 above)
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:27]: You are right, my latest commit fixes this by introducing a slightly complicated bootstrapping logic to allow pex builder to be either a py_binary from rules_python with a pip dependency (for local dev only) or a py_binary from rules_python with no external dependencies but a  pex builder packaged into a pex zip to prevent a python dependency.

Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:27]: > Ideally we wouldn't have a Python dependency at all, is there another way to interact with the PEX library, can we vendor the dependency?

vendoring is another approach we could take, it's open for a debate.

File: py/private/py_pex.bzl [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+def _runfiles_path(file, workspace):
+    if file.short_path.startswith("../"):
+        return file.short_path[3:]
+    else:
+        return workspace + "/" + file.short_path
+
+def _map_srcs(f, workspace):
+    site_packages_i = f.path.find("site-packages")
+    # determine if the src if a third party.
+    if site_packages_i != -1 and f.path.count("/",  site_packages_i) == 2:
```
[the remark this answers is not in the document]
Comment (thesayyn) [reply py/private/py_pex.bzl]: I will explain this in detail once it is close to the finish line. i am not sure if this is the way to install third-party deps yet. I'll wait for some feedback from some internal users first.

File: py/tools/pex/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    name = "use_dev_pex",
+    build_setting_default = False,
+)
+
+config_setting(
+    name = "should_use_dev_pex",
+    flag_values = {
+        ":use_dev_pex": "true",
+    },
+)
```
Comment (mattem) [inline py/tools/pex/BUILD.bazel:62]: There's an existing mechanism to select between the release and dev version of tools by switching on the `IS_PRERELEASE` symbol in `tools/version.bzl`, can we hook into the same structure here, rather than adding something new?

File: py/tools/pex/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,71 @@
+load("@aspect_bazel_lib//lib:write_source_files.bzl", "write_source_file")
+load("@bazel_skylib//rules:common_settings.bzl", "bool_flag")
+load("@rules_python//python:defs.bzl", "py_binary")
+load("//py:defs.bzl", _py_binary = "py_binary")
```
Comment (mattem) [inline py/tools/pex/BUILD.bazel:4]: Perhaps rename one symbol explicitly to `rules_python_py_binary`? It's hard to follow with just the _ rename.

File: py/tools/pex/BUILD.bazel (hunk 4 above)
Comment (mattem) [reply py/tools/pex/BUILD.bazel:17]: Given that pex is a single wheel with no dependencies, I'm not clear why we need to bootstrap it via a zip? 
Could we instead fetch the wheel via an `htt_file`, have rules_py expand it with the existing tools, then depend on that? rules_py has a fetch dependencies function where it declares other rule dependencies, so we could add the `http_file` into there.

Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:17]: Yeah, that sounds better, i'll do that once get back to this again.

File: py/tools/pex/BUILD.bazel (hunk 6 above)
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:62]: with the http_file approach, we don't need this as the pex builder always will be the same target.

**Comment by dizzy57** [comment]
I was unable to make it work with bzlmod, here's a minimal repro: https://github.com/dizzy57/bazel_py_minimal

File: py/tools/pex/BUILD.bazel (hunk 4 above)
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel]: This is now fixed.

File: examples/py_pex_binary/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        "TEST": "1"
+    },
+    deps = [
+        "@pypi_cowsay//:pkg",
+        "@bazel_tools//tools/python/runfiles",
+    ],
+)
+
+
+
```
Comment (mattem) [inline examples/py_pex_binary/BUILD.bazel:18]: Can these be tidied? I think I've commented on empty space in this diff before, perhaps a pass over would be good.

File: py/private/py_pex_binary.bzl [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    args = ctx.actions.args()
+
+    # Copy workspace name here to prevent ctx
+    # being transferred to the execution phase.
+    workspace_name = str(ctx.workspace_name)
+
+    args.add_all(
+        ctx.attr.inject_env.items(), 
+        map_each = lambda e: "--inject-env=%s=%s" % (e[0], e[1]),
+        allow_closure = True,
```
Comment (mattem) [inline py/private/py_pex_binary.bzl:59]: Does this need to capture the closure? Seems like it only references the var in the lambda.

File: py/private/py_pex_binary.bzl [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    ctx.actions.run(
+        executable = ctx.executable._pex,
+        inputs = runfiles.files,
+        arguments = [args],
+        outputs = [output],
+        mnemonic = "PyPex",
+        progress_message = "Building PEX binary %{label}",
+        # Unfortunately there is no way to disable pex cache, so just set it to . allow
+        # bazel to discard cache once the action is done. 
+        # TODO: this is probably not the right thing to do if the action is unsandboxed.
```
Comment (mattem) [inline py/private/py_pex_binary.bzl:87]: What is the right thing to do? Should we create an issue rather than an TODO in the code that will never be todone?

File: py/repositories.bzl [hunk 11]
```diff
@@ -42,4 +53,40 @@ def rules_py_dependencies():
...
+    Args:
+        name: prefix used in created repositories
+        register: whether to call the register_toolchains, should be True for WORKSPACE and False for bzlmod.
+        is_prerelease: True iff there are no pre-built tool binaries for this version of rules_py
+    """
+
+    # the url and digest information can be found at https://pypi.org/pypi/pex/json
+    # WARNING: when updated, also update MODULE.bazel
+    http_file(
+        name = "pex_2_3_1",
```
Comment (mattem) [inline py/repositories.bzl:73]: Do you think we should namespace this more? `rules_py_pex_231` for example?

File: py/repositories.bzl [hunk 12]
```diff
@@ -42,4 +53,40 @@ def rules_py_dependencies():
...
+def rules_py_toolchains(name = DEFAULT_TOOLS_REPOSITORY, register = True, is_prerelease = IS_PRERELEASE):
+    """Create a downloaded toolchain for every tool under every supported platform.
+
+    Args:
+        name: prefix used in created repositories
+        register: whether to call the register_toolchains, should be True for WORKSPACE and False for bzlmod.
+        is_prerelease: True iff there are no pre-built tool binaries for this version of rules_py
+    """
+
+    # the url and digest information can be found at https://pypi.org/pypi/pex/json
```
Comment (mattem) [inline py/repositories.bzl:70]:
```suggestion
    # The url and digest information can be found at https://pypi.org/pypi/pex/json
```

File: py/private/py_pex_binary.bzl (hunk 10 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:87]: i must address this before landing this PR.

File: py/private/py_pex_binary.bzl (hunk 9 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:59]: nope, left over. will remove

**Comment by thesayyn** [comment]
Also, CPython=={major}.{minor}.* should be the default constraint.

File: py/private/py_pex_binary.bzl (hunk 10 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl]: done

**Comment by thesayyn** [comment]
@mattem I have removed some unrelated diff from my failed attempt before using PEX.  Now diff should be smaller and easier to understand.

File: py/private/py_pex_binary.bzl [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        DefaultInfo(files = depset([output]), executable = output),
+        # See: https://github.com/bazelbuild/bazel/blob/b4ab259fe1cba8a108f1dd30067ee357c7198509/src/main/starlark/builtins_bzl/common/python/py_executable_bazel.bzl#L265
+        OutputGroupInfo(
+            python_zip_file = depset([output])
+        )
+    ]
+
+
+_attrs = dict({
+    "binary": attr.label(executable = True, cfg = "target"),
```
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:123]:
```suggestion
    "binary": attr.label(executable = True, mandatory = True, cfg = "target"),
```
Building a pex without a `binary` doesn't make much sense.

Comment (njlr) [reply py/private/py_pex_binary.bzl:123]: I think that a PEX can be used as an environment only - IIUIR the entry point(s) are option in the PEX tool https://docs.pex-tool.org/buildingpex.html#specifying-entry-points

File: py/private/py_pex_binary.bzl [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+def _runfiles_path(file, workspace):
+    if file.short_path.startswith("../"):
+        return file.short_path[3:]
+    else:
+        return workspace + "/" + file.short_path
+
+exclude_paths = [
+    "toolchain",
+    "aspect_rules_py/py/tools/",
+    "rules_python~~python~",
```
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:16]: > `rules_python~~python~`

Will this path change if we set `--incompatible_use_plus_in_repo_names`? It's going to be set to true by default in bazel 8.

context: https://github.com/bazelbuild/bazel/issues/23127

File: py/private/py_pex_binary.bzl (hunk 13 above)
Comment (dizzy57) [reply py/private/py_pex_binary.bzl:123]: @njlr there is no `deps` attribute in this rule (and I don't think we need it), so setting `binary` is the only way to specify what goes into PEX. In any way, the rule currently fails if you don't specify `binary` attribute:
```
ERROR: [...]/cli_tool/BUILD.bazel:19:14: in py_pex_binary rule //cli_tool:empty_pex:
Traceback (most recent call last):
	File "[...]/external/aspect_rules_py~/py/private/py_pex_binary.bzl", line 61, column 22, in _py_python_pex_impl
		runfiles = binary[DefaultInfo].data_runfiles
Error: type 'NoneType' has no operator [](Provider)
```

File: py/private/py_pex_binary.bzl (hunk 14 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:16]: Yes, this will stop working.

Comment (thesayyn) [reply py/private/py_pex_binary.bzl:16]: Sent a filter for plus version as well. LMK if that does it.

File: py/private/py_pex_binary.bzl (hunk 13 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:123]: done

File: py/private/py_pex_binary.bzl (hunk 9 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:76]: Please see the comment below the argument. 

https://github.com/aspect-build/rules_py/pull/324/files#diff-bdbfb557a8b870d00314cf49db0fad2611b37d9d20d6c055034aef05b30a6e65R77

File: py/private/py_pex_binary.bzl [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    "rules_python++python+",
+    "aspect_rules_py~/py/tools/",
```
[the remark this answers is not in the document]
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:18]: It changes based on `--incompatible_use_plus_in_repo_names` flag. Left a comment above them.

File: py/private/py_pex_binary.bzl (hunk 9 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: > # this is needed to allow passing a lambda (with workspace_name) to map_each

Comment (mattem) [reply py/private/py_pex_binary.bzl:80]: But this isn't that call, the `map_each` here is `lambda e: "--inject-env=%s=%s" % (e[0], e[1]),` so how is the call below related?

Comment (dizzy57) [reply py/private/py_pex_binary.bzl:80]: Lambda for `inject_env`'s `map_each` does not contain any references to the `workspace_name`.

Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: i could get rid of it by creating a global `def`, this feels more localized and i did not see any performance degradation.

File: py/private/py_pex_binary.bzl [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+        outputs = [output],
+        mnemonic = "PyPex",
+        progress_message = "Building PEX binary %{label}",
+    )
+
+    return [
+        DefaultInfo(files = depset([output]), executable = output),
+        # See: https://github.com/bazelbuild/bazel/blob/b4ab259fe1cba8a108f1dd30067ee357c7198509/src/main/starlark/builtins_bzl/common/python/py_executable_bazel.bzl#L265
+        OutputGroupInfo(
+            python_zip_file = depset([output])
```
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:122]: Do we need to return `python_zip_file` output group here?

Looks like `py_pex_binary` is a rule on it's own  by now and does not need any output groups. I think it's fine to return just the `DefaultInfo` provider.

File: py/private/py_pex_binary.bzl (hunk 9 above)
Comment (dizzy57) [reply py/private/py_pex_binary.bzl:80]: https://bazel.build/rules/lib/builtins/Args#parameters_1 says
> To avoid unintended retention of large analysis-phase data structures into the execution phase, the map_each function must be declared by a top-level def statement; [...]

so I think it's better to rewrite it as a global `def`

Comment (mattem) [reply py/private/py_pex_binary.bzl:80]: In that case, can the comment be updated, as it's unrelated to what is happening in his call.

File: py/private/py_pex_binary.bzl (hunk 16 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:122]: you are right, it's not necessary.

File: py/private/py_pex_binary.bzl (hunk 9 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: fixed the comment.

File: py/private/py_pex_binary.bzl (hunk 16 above)
Comment (thesayyn) [reply py/private/py_pex_binary.bzl]: removed.

**Review by mattem — APPROVED** [review]
